srai/neighbourhoods/neighbourhood.py

```diff
@@ -0,0 +1,101 @@
+    `get_neighbours_up_to_distance` and `get_neighbours_at_distance` methods for performance
+    reasons.
+    See the `H3Neighbourhood` class for an example.
+    """
+
+    @abstractmethod
+    def get_neighbours(self, index: IndexType) -> Set[IndexType]:
+        """
+        Get the direct neighbours of a region using its index.
+
+        Args:
+            index (IndexType): Unique identifier of the region.
+                Dependant on the implementation.
+
+        Returns:
+            Set[IndexType]: Indexes of the neighbours.
+        """
+        pass  # pragma: no cover
+
+    def get_neighbours_up_to_distance(self, index: IndexType, distance: int) -> Set[IndexType]:
+        """
+        Get the neighbours of a region up to a certain distance.
+
+        Args:
+            index (IndexType): Unique identifier of the region.
+                Dependant on the implementation.
+            distance (int): Maximum distance to the neighbours.
+
+        Returns:
+            Set[IndexType]: Indexes of the neighbours.
+        """
+        neighbours_with_distances = self._get_neighbours_with_distances(index, distance)
+        neighbours: Set[IndexType] = seq(neighbours_with_distances).map(lambda x: x[0]).to_set()
+        neighbours.discard(index)
+        return neighbours
+
+    def get_neighbours_at_distance(self, index: IndexType, distance: int) -> Set[IndexType]:
+        """
+        Get the neighbours of a region at a certain distance.
+
+        Args:
+            index (IndexType): Unique identifier of the region.
+                Dependant on the implementation.
+            distance (int): Distance to the neighbours.
+
+        Returns:
+            Set[IndexType]: Indexes of the neighbours.
+        """
+        neighbours_up_to_distance = self._get_neighbours_with_distances(index, distance)
+        neighbours_at_distance: Set[IndexType] = (
+            seq(neighbours_up_to_distance)
+            .filter(lambda x: x[1] == distance)
+            .map(lambda x: x[0])
+            .to_set()
+        )
+        neighbours_at_distance.discard(index)
+        return neighbours_at_distance
+
+    def _get_neighbours_with_distances(
+        self, index: IndexType, distance: int
+    ) -> Set[Tuple[IndexType, int]]:
+        visited = set()
+        visited_with_distances = set()
+        to_visit: Queue[Tuple[IndexType, int]] = Queue()
+
+        to_visit.put((index, 0))
+
+        while not to_visit.empty():
+            current, current_distance = to_visit.get()
+
+            visited.add(current)
+            visited_with_distances.add((current, current_distance))
+            if current_distance < distance:
+                current_neighbours = self.get_neighbours(current)
+                for neighbour in current_neighbours:
+                    if neighbour not in visited:
+                        to_visit.put((neighbour, current_distance + 1))
+
+        return visited_with_distances
```

Comment:
wouldn't it be better to replace it with a `dict`?
```suggestion    
        visited_indexes = dict()    
        to_visit: Queue[Tuple[IndexType, int]] = Queue()

        to_visit.put((index, 0))

        while not to_visit.empty():
            current_index, current_distance = to_visit.get()

            visited_indexes[current_index] = current_distance
            if current_distance < distance:
                current_neighbours = self.get_neighbours(current_index)
                for neighbour in current_neighbours:
                    if neighbour not in visited_indexes:
                        to_visit.put((neighbour, current_distance + 1))

        return set(visited_indexes.items())
```